docs/docs_build.py

```diff
@@ -0,0 +1,171 @@
+from __future__ import annotations
+from contextlib import contextmanager
+from functools import cached_property
+import json
+from .logging import Logger
+import os
+import shutil
+from pathlib import Path
+import re
+from typing import TYPE_CHECKING, Generator, List, Optional, cast
+from io import BytesIO
+import zipfile
+
+import requests
+
+from docs.prepare_prior_versions import prepare_prior_versions
+
+if TYPE_CHECKING:
+    from invoke.context import Context
+
+
+class DocsBuilder:
+    def __init__(
+        self,
+        context: Context,
+        current_directory: Path,
+        is_pull_request: bool,
+        is_local: bool,
+    ) -> None:
+        self._context = context
+        self._current_directory = current_directory
+        self._is_pull_request = is_pull_request
+        self._is_local = is_local
+
+        self._current_commit = self._run_and_get_output("git rev-parse HEAD")
+        self._current_branch = self._run_and_get_output(
+            "git rev-parse --abbrev-ref HEAD"
+        )
+
+    def build_docs(self) -> None:
+        """Build API docs + docusaurus docs.
+        Currently used in our netlify pipeline.
+        """
+        self._prepare()
+        self.logger.print_header("Building docusaurus docs...")
+        self._context.run("yarn build")
+
+    def build_docs_locally(self) -> None:
+        """Serv docs locally."""
+        self._prepare()
+        self.logger.print_header("Running yarn start to serve docs locally...")
+        self._context.run("yarn start")
+
+    def _prepare(self) -> None:
+        """A whole bunch of common work we need"""
+        self.logger.print_header("Preparing to build docs...")
+        self._load_files()
+
+        self.logger.print_header(
+            "Updating versioned code and docs via prepare_prior_versions.py..."
+        )
+        # TODO: none of this messing with current directory stuff
+        os.chdir("..")
+        prepare_prior_versions()
+        os.chdir("docusaurus")
+        self.logger.print("Updated versioned code and docs")
+
+        self._invoke_api_docs()
+        self._checkout_correct_branch()
+
+    @contextmanager
+    def _load_zip(self, url: str) -> Generator[zipfile.ZipFile, None, None]:
+        response = requests.get(url)
+        zip_data = BytesIO(response.content)
+        with zipfile.ZipFile(zip_data, "r") as zip_ref:
+            yield zip_ref
+
+    def _load_files(self) -> None:
+        """Load oss_docs_versions zip and relevant versions from github.
+        oss_docs_versions contains the versioned docs to be used later by prepare_prior_versions, as well
```

Comment:
Add a blank space between initial summary line of docstring and rest of docstring: https://google.github.io/styleguide/pyguide.html#381-docstrings